### ISSUE TITLE
Create a column summary for legacy sprint project in python

### ISSUE THREAD

**Issue by mikeaintworkin** [thisaintwork/iqss_gh_reporting#5]
Create a column summary using pandas on data returned from the legacy sprint project.
Display the data in the same way that we measure it for sprints.

Note: 
If we trim the closed issues column down from around 500 to 100 or so this will run fairly quickly.
Right now it takes around 3 minutes to run.
We can trim that column by removing older closed issues from the project.

context:
This is part of the closeout where I get all the code into python.
This functionality will be needed as a bridge between now and when we transition to projectV2 objects for the sprint.

**Comment by mikeaintworkin** [comment]
Going into this issue I already had working python code to read the legacy project and print out the columns and also save the columns to a flat file.

Today I
- learned some pandas using chatgpt and the docs
- converted the code to use pandas 
- cleaned up sloppy code around forming the output file path and name
- did a little experimenting with pandas - like how hard would it be to ... export to xml. wow. easy
- thought through the next step logic on paper.

Step 1
- Create columns 
  -  date
  - sprint name
  - InTheSprint 
  - ready for sprint
  -  This sprint
  - WIP
  - Ready for review
  - Review
  - Ready for QA
  - QA

Assign sizes to PRs
- If this is anytime after the initial kickoff count (which happens only once) 
  - Find the issues that are tied to any PRs that do not have a size and use that size

Create InTheSprint count
- Sum up:
    - Sum of these columns
      -  This sprint
      - WIP
      - Ready for review
      - Review

Create the rest of the sums
  - ready for sprint
  -  This sprint
  - WIP
  - Ready for review
  - Review
  - Ready for QA
  - QA

Output the results as a flat file

**Comment by mikeaintworkin** [comment]
I have the code already for summarizing each column.
It will not be hard to get from there to summarizing accross a few columns.
The new code will be the tracking down of the attached PRs.


This will have to be dealt with even when the transition is made to the GitHub projectV2 so it's not wasted work.

My thought is that in graphql I can get closingIssuesReferences from a pullrequest

**Comment by mikeaintworkin** [comment]
I left off working on SprintSummaryFrame.
I left off in: _add_size_column
 debugging.
-I'm using old code and I am running into dumb column header problems.
-I'm thinking maybe rather than keep hitting every reference I might define a tuple with all the column names and use them that way. dunno. getting late heading home.

Code that is checked in works.
This code is not yet checked in because - well - it doesn't.

**Comment by mikeaintworkin** [comment]
I won't be able to look at this again until tomorrow evening.

**Comment by mikeaintworkin** [comment]
I'm going to pass in the headers I care about as a list

**Comment by mikeaintworkin** [comment]
Basic hacky code is checked in.
produces this:
```
370	▶ SPRINT READY
109	This Sprint 🏃‍♀️ 🏃 
146	IQSS Team - In Progress  💻
56	Ready for Review ⏩
33	Review 🔎
12	Ready for QA ⏩
20	QA ✅
897	Done 🚀
344	ActiveSprint
```

Does not include the lookup to get the points associated with issues that are off the board, replaced by unsized PRs.

**Comment by mikeaintworkin** [comment]
I did one last ChatGPT query tonight.
I want to see if I can cut down on the query time if I include a filter somehow.
ChatGPT says  i can do it. I'll try it.

```
from datetime import datetime, timedelta
from github import Github

# create a GitHub API object and authenticate with your access token
g = Github("<access-token>")

# specify the owner and repository name for the project
owner = "your-username"
repo_name = "your-repo-name"

# get the repository object
repo = g.get_repo(f"{owner}/{repo_name}")

# specify the query to retrieve cards that have issues or pull requests attached that are currently open or were closed less than a year ago
query = f"repo:{owner}/{repo_name} is:open is:issue is:pr updated:>= {datetime.now() - timedelta(days=365)}.strftime('%Y-%m-%d')"

# search for issues or pull requests that match the query
search_results = g.search_issues(query=query)

# loop through each issue or pull request and print the cards that have them attached
for issue_or_pr in search_results:
    cards = issue_or_pr.get_cards()
    for card in cards:
        # print the card's information
        print(f"{card.note} - {issue_or_pr.html_url}")
``

**Comment by mikeaintworkin** [comment]
Technically I think I've actually finished this issue and should open another.

**Comment by mikeaintworkin** [comment]
I realized that I have one more piece to do here.
I need to output the results to a text file.

**Comment by mikeaintworkin** [comment]
I went far afield from simply getting the printing to work.
Today 
- I took a side trip in the TDD and learned the basics.
- I thought through the workflow a bit.
- Realized that I don't want objects that do everything.
- experimented with how I would deal with the needed consitency in column names.
- Played with querying GitHub legacy project for only closed issues last touched less than 3 months ago.  I still ended up with a loop. I don't want to change how I go into the api. org->project->cards->issue|pr and it seems like the cards object has no way to  query it using PyGithub that allows you to limit things at the server side.
- I want to do more TDD but I don't want to spend all my time figuring out how to setup testing for the API related stuff so I will stick to things where I'm working with a local file.



![Image](https://user-images.githubusercontent.com/123129351/233879718-1ecda510-f7cb-4a2b-a763-622eab6eccc2.png)

Next:
- [ ] Make the output saved from the summary save in exactly the format required here.
- [ ] Add the logic to track down the pull requests.

**Comment by mikeaintworkin** [comment]
Today:
- I ran a mostly succesful workflow using: mn-snapshot_sprint_from_api.py
- Worked on SprintSummaryFrame  so that it returns (dummy) data for the additional columns that I want.
- Ran into issues that I'm trying to track down.
- To make that easier I moved over to a new worfklow mn-snapshot_sprint_from_file.py
- I'm reading in the results of the mn-snapshot_sprint_from_api.py workflow.
- This way I can avoid calling the API repeetedley while working through this issue.
- I haven't figured out how I want to to replace the dummy data for date and sprint in the summary report. That can wait until after I get this bug fixed and the pulll request logic figured out.
- The whole concept of how the python project is setup clicked today. I broke out my workflows into a separate python package so that things are much neater now.
- I may further break out the things I'm using as utility functions from the things I have as objects but that's really low priority.
- Starting to feel at home in pycharm and python. wahoo.

next steps:
- [ ] step backwards & complete the debug of SprintSummaryFrame using mn-snapshot_sprint_from_file.py
- [ ] Finish the previous next steps


The flow using the API right now.
```
    df = pdio.LegacyProjectCards(
        access_token=auth_token_val,
        organization_name=args.organization_name,
        project_name=args.proj_name)
    df.fetch_data()
    df.print_project_cards()
    pdio.write_dataframe(df=df.dataframe()) # raw data
    df.print_project_cards()
    dfsum = pdio.SprintSummaryFrame(df_in=df.dataframe())
```

The flow using an input file right now
```
    df = pdio.DFFromFile(
        in_dir=args.dir_name,
        file_name=args.filename)
    dfsum = pdio.SprintSummaryFrame(df_in=df.dataframe())
```

**Comment by mikeaintworkin** [comment]
I'm going to close this.
I was able to run this workflow tonight:
```
    df = pdio.DFFromFile(
        in_dir=args.dir_name,
        file_name=args.file_name)
    dfSzer = pdio.SprintCardSizer(df_in=df.dataframe())
    df =dfSzer.dataframe()
    pdio.write_dataframe(df)
    spsumrzr = pdio.SprintSizeSummarizer(df_in=df)
    print(spsumrzr.sprint_summary_line())

```

and get this result:
```
Column	Size
▶ SPRINT READY	370
This Sprint 🏃‍♀️ 🏃	46
IQSS Team - In Progress  💻	159
Ready for Review ⏩	53
In Review 🔎	36
QA ✅	6
Done 🚀	623
ActiveSprint	294

▶ SPRINT READY This Sprint 🏃‍♀️ 🏃 IQSS Team - In Progress  💻 Ready for Review ⏩ In Review 🔎 QA ✅ Done 🚀 ActiveSprint      DateTime  SprintName
           370                 46                        159                 53          36    6    623          294 datetimestamp sprint_name
```

**Comment by mikeaintworkin** [comment]
I'm breaking out the logic to add the pull requests to a new issue.
